### DISCUSSION TITLE
PVP Sonar marker Icons

### DISCUSSION THREAD

**Discussion by Leonidas-ai** [FakeFishGames/Barotrauma#4844] [Ideas and suggestions]
Would it be possible to enable sonar icons in PVP through a command or toggle? The new map layouts are much more interesting but they can lead to a significantly higher difficulty in finding enemy subs. Even when sonar icons are enabled in PVP client side, I have had problems finding enemy subs because of how well the new cave systems are created so I think other players could benefit from it too.  Perhaps simply a minimum range before sonar markers are disabled could be used (which would also help greatly with the clutter from sonar markers).

**Comment by NilanthAnimosus** [comment]
Make this a server option, and instead of direct-to-submarine icons, use some mechanism to report if the submarine has entered or left a specific range. such as a mission objective for high alert within X meters and outside of X meters (Such as say, 500? 300?)

This would not take too much tweaking and should be configurable server by server, as I can picture some not wanting this, but the issue is also legitimately there and some others may want to alleviate this understandably. knowing exactly where and when the enemy submarine will be however may easily detract from sub vs sub though.

If wanting to use and stick with icons however, make them directional only. as to know the submarine is right or left in a delayed way that prevents narrowing it down too fast. some may want to hide their submarine in waiting to ambush as the others crosses for example to look like a wreck or part of the level geometry.

**Reply by Leonidas-ai** [reply]
Oh wow. I've always wanted to ask you about your Nilmod but I saw how complex it is and how difficult it would be to maintain. I've been told there's no way for the server to change "submarine.ShowSonarMarker" to make it available client side and I have only had success with making it visible for myself in pvp. Are you say that if I edited "public override void Update(float deltaTime)" in "combatmission.cs", it's possible the server could send mission objective updates that could alter a client's HUD, GUI, or sonar display?

**Reply by NilanthAnimosus** [reply]
I wrote my own code for my own custom icons in that mod, It was not simply done by flicking a switch I am afraid (And the clients would have the mod too even in my case, the XML settings to display such synced with those who had those edits). you could technically achieve anything by editing the code, except for a client sided change you most likely would need to edit their clients, I cannot think of a good clean way to 'fake' a sonar target client side either - as most of the sonar information is setup client side at level generation usually or handled client side by specific items.

Changes like these in theory, typically take edits on both sides. Unless there already exists some form of networking code and switches to inform clients to toggle this behavior. If its a console command, you would have an easier time sending the final result of the effects of that console command to everybody, but I am not up to date what exists or what that particular section of code looks like - just tossing my thoughts there if you were considering modding it.

**Reply by Leonidas-ai** [reply]
So ShowSonarMarker is a simple boolean that would activate everything I want clientside if I could access it. However, it is part of the shared source code rather than strictly sever side. I see no reason why this couldn't be server side. Would you suggest I make a simple pull request that moves this boolean server side so the server can control it? It appears to be baked into the level on load and I haven't had any success changing it while a match is in progress (client or server side). Do you know how a server might be able to access the level information before loading a match so I could alter it?

**Reply by NilanthAnimosus** [reply]
Its really just a case of them adding syncing for that particular variable, but the amount of data in a packet to send is considered a premium, as there is a cost to syncing things so generally they try their best not to sync additional unnecessary data, they generally go off syncing only what needs to be sent.

Its not a case of it being client/shared/server code. But more if the server is to dictate a feature being enabled or not, it must communicate this to the clients, handling the drawing and if it is allowed for a particular client to draw it still should be a clientside value (or shared) and set from the server by the networking, wherever that would be handled in this case (Start game coroutine, a lobby setting to sync, there is a good few places, even the events and information for the mission type itself).

Any feature could be implemented in theory, so where it is or how its setup shouldn't matter too much beyond how feasible it may be, and I agree some form of this could definitely be nice to have - I have definitely spent a few games going right past the enemy submarine and just going back and forth up and down for a good 40 minutes before finding out they died anyways.

Exposing this to say, code modding would be a fairly odd thing as it sounds like it should just be a feature in the XML definitions and perhaps a mod that way to show the icons (such as a simple steam workshop XML override mod), at least if your talking about exposing this and changing the code for modding support.

If wanting this as a feature however and not thinking of modding, it would be better to consider the implementation and how it would be designed in a game play sense that would not be considered a bad feature for various PvP server hosts (As not everyone will agree on wanting this, and those that do will have differing ideas on how it should be done, such as myself thinking it would be bad for pin-point accuracy to the enemy submarine over a generalized guide or notification).

**Comment by Leonidas-ai** [comment]
Also, this is the bit of code that would need to move from shared to server. And probably should be moved anyway to prevent cheating.

```
            foreach (Submarine submarine in Submarine.Loaded)
            {
                //hide all subs from sonar to make sneak attacks possible
                submarine.ShowSonarMarker = true;
            }
```
